src/Toolbox/Util/Graph/BFS.hs

```diff
@@ -0,0 +1,134 @@
+{-# LANGUAGE BangPatterns #-}
+
+-- | Breadth-First-Search (BFS)
+module Toolbox.Util.Graph.BFS
+  ( -- * internal state for BFS
+    BFSState (..),
+    BFSPath (..),
+    MultiBFSPath (..),
+
+    -- * single-seed use case
+    stepBFS,
+    runStagedBFS,
+
+    -- * multi-seed use case
+    stepMultiseedBFS,
+    runMultiseedStagedBFS,
+  )
+where
+
+import Control.Monad (join)
+import Control.Monad.Extra (loopM)
+-- import Control.Monad.IO.Class (liftIO)
+import Control.Monad.Trans.State
+  ( StateT (..),
+    evalStateT,
+    get,
+    modify',
+  )
+import Data.Bifunctor (bimap)
+import Data.Either (partitionEithers)
+import qualified Data.Foldable as F
+import Data.IntMap (IntMap)
+import qualified Data.IntMap as IM
+import Data.IntSet (IntSet)
+import qualified Data.IntSet as IS
+import Data.Maybe (maybeToList)
+
+newtype BFSState = BFSState
+  { bfsVisited :: IntSet
+  }
+  deriving (Show)
+
+data BFSPath = BFSPath
+  { bfsSearchResult :: ![[Int]]
+  , bfsNextStage :: ![Int]
+  }
+  deriving (Show)
+
+stepBFS ::
+  forall m.
+  (Monad m) =>
+  IntMap [Int] ->
```

Comment:
here is your ever-growing queue, huh?

Reply:
Ah. no. this `IntMap [Int]` is the graph itself. I use node-to-outgoing-node(i.e. edge) map.
The queue is represented with `IntSet` in BFSState, and individual paths are [[Int]], where the nested list is used for splitting each search stage.